Fix integer division of ActiveSupport::Duration

aarongough opened the pull request:
### Summary

Fixes issue: https://github.com/rails/rails/issues/33179

In the case where an `ActiveSupport::Duration` object was divided by an integer, it would often return `0 seconds` as the result incorrectly. This is because it was performing integer division on the `#parts` of the original duration. So 1 year divided by 2 would result in zero:

```ruby
1.year / 2
 => 0 seconds
```

The `#value` of the resulting duration was correct, but the `#parts` was not and this would cause issues in any case that used the parts like when performing math against a date or time object, eg:

```ruby
> Time.new(2018, 1, 1, 0, 0, 0, "+00:00") - (1.day / 24)
=> 2018-01-01 00:00:00 +0000
# should be 2017-12-31 23:00:00 +0000
```

sikachu reviewed (approved):
Changes look good.

Is this also happening on `master` as well? If so, this PR probably should target `master` then the core team will decide if it should be backported to 5.2.x.

bogdanvlviv commented:
@aarongough Thank you for the Pull Request!

1) Could you please change the target branch from `5-2-stable` to `master` since there is the same issue, then, as @sikachu  mentioned,  the core team will decide if it should be backported to `5-2-stable`.

2) I also see the same issue with divide duration by `Scalar`
```ruby
(1.week / ActiveSupport::Duration::Scalar.new(7)).parts # => {}`
``` 

Could you please apply this diff too?
```diff
diff --git a/activesupport/lib/active_support/duration.rb b/activesupport/lib/active_support/duration.rb
index 88897f811e..6392ca8c3e 100644
--- a/activesupport/lib/active_support/duration.rb
+++ b/activesupport/lib/active_support/duration.rb
@@ -266,7 +266,7 @@ def *(other)
     # Divides this Duration by a Numeric and returns a new Duration.
     def /(other)
       if Scalar === other
-        Duration.new(value / other.value, parts.map { |type, number| [type, number / other.value] })
+        Duration.build(value / other.value)
       elsif Duration === other
         value / other.value
       elsif Numeric === other
diff --git a/activesupport/test/core_ext/duration_test.rb b/activesupport/test/core_ext/duration_test.rb
index 240ae3bde0..5490cee4aa 100644
--- a/activesupport/test/core_ext/duration_test.rb
+++ b/activesupport/test/core_ext/duration_test.rb
@@ -461,6 +461,17 @@ def test_scalar_divide
     assert_equal "no implicit conversion of String into ActiveSupport::Duration::Scalar", exception.message
   end

+  def test_divide_by_scalar_parts
+    assert_equal({ days: 1 }, (1.year / ActiveSupport::Duration::Scalar.new(365.2425)).parts)
+    assert_equal({ hours: 1 }, (1.day / ActiveSupport::Duration::Scalar.new(24)).parts)
+    assert_equal({ minutes: 1 }, (1.hour / ActiveSupport::Duration::Scalar.new(60)).parts)
+    assert_equal({ seconds: 1 }, (1.minute / ActiveSupport::Duration::Scalar.new(60)).parts)
+  end
+
+  def test_date_minused_by_divided_duration_by_scalar
+    assert_equal Date.civil(2017, 12, 31), Date.civil(2018, 1, 1) - (1.week / ActiveSupport::Duration::Scalar.new(7))
+  end
+
   def test_scalar_modulo
     scalar = ActiveSupport::Duration::Scalar.new(10)
```

3) Could you squash your commits into one?

eugeneius commented:
A duration's parts describe an exact time interval; its value is an approximation of that interval in seconds. Throwing away the parts and rebuilding them from the value loses information, and will cause incorrect behaviour when the interval is converted to or from a unit of variable length.

On master:

```ruby
$ tools/console
irb(main):001:0> 90.days
=> 90 days
irb(main):002:0> 90.days / 2
=> 45 days
irb(main):003:0> Time.now.beginning_of_year + 90.days
=> 2018-04-01 00:00:00 +0100
irb(main):004:0> Time.now.beginning_of_year + (90.days / 2)
=> 2018-02-15 00:00:00 +0000
```

On this branch:

```ruby
$ tools/console
irb(main):001:0> 90.days
=> 90 days
irb(main):002:0> 90.days / 2
=> 1 month, 2 weeks, 13 hours, 30 minutes, and 54.0 seconds
irb(main):003:0> Time.now.beginning_of_year + 90.days
=> 2018-04-01 00:00:00 +0100
irb(main):004:0> Time.now.beginning_of_year + (90.days / 2)
=> 2018-02-15 13:30:54 +0000
```

bogdanvlviv commented:
@eugeneius Thank you for the clarification.
Would be great to add a test that describes this behavior to prevent any regression in the future.

eugeneius commented:
Good idea @bogdanvlviv - I've opened https://github.com/rails/rails/pull/33217 to add the missing test coverage.

aarongough commented:
Thanks for the feedback guys! I have rebased against master, and squashed everything into one commit.

At the moment the only alternative implementation I can think of is to force any division of the `parts` to be a floating point operation rather than allowing integer math. I will update the PR with an implementation of this so you can give your feedback.

I used the updated tests from @eugeneius to validate this locally and it was green.

bogdanvlviv commented on activesupport/test/core_ext/duration_test.rb:
```diff
@@ -158,6 +158,14 @@ def test_date_added_with_multiplied_duration
   end
 
+  def test_date_added_with_divided_duration
+    assert_equal Date.civil(2017, 1, 3), Date.civil(2017, 1, 1) + 4.days / 2
```
I would change this to
```ruby
assert_equal Date.civil(2017, 12, 31), Date.civil(2018, 1, 1) - (1.week / 7)
```
since `4.days / 2` works well with and without the changes.
See #33179

aarongough replied:
Hmmm, not sure where that came from as I know my original was `1.week / 7` and was minused not added! I must have discarded the wrong chunk when composing this PR. Will fix!

I will also add some more targeted tests as well regarding `parts`.

aarongough commented:
@bogdanvlviv @eugeneius I added several new test cases that codify the existing behaviour of `parts` in relation to maintaining precision, as well as failing test cases that expose the integer and scalar division issues. Let me know what you think!

bogdanvlviv reviewed (approved):
LGTM! Thank you for working on this.

aarongough commented:
Excellent! Thanks for the help @bogdanvlviv!

aarongough commented:
@eugeneius Thanks for the feedback. I understand that the change in the behaviour of `Duration` was implemented in order to fix confusion regarding how it interacted with `Date`, but I fundamentally don't believe that anyone expects this:

```ruby
> 1.day / 2
=> 0 seconds
```

`Duration` should be a reasonable abstraction for dealing with time spans, and no-one expects a single day or year to be indivisible, which is how it now behaves.

I think the behaviour of `ActiveSupport::Duration` should be rolled back to how it behaved prior to `5.1.0`, with numeric operations on `Duration` acting upon the seconds `#value` of the duration. The new behaviour breaks existing logic without raising any exceptions and I'm sure it's going to cause difficult-to-trace issues in a lot of projects as people upgrade.

I will update my branch to remove my implementation and just leave the failing test cases, in case anyone else has a good idea for how this can be fixed! Thanks for your help so far!

@jeremy @cornerfade @jfilipe

aarongough committed:
Fix integer division of duration.

aarongough committed:
Remove fix implementation, leave failing test cases.
